File: 03-week-3--additional-practice/04-bonus-5--consecutive-substrings/ruby/consecutive_substrings.rb
def consecutive_substrings(string)
  # type your code in here

  subs = []   # create a variable with an empty array to store substrings

  string.split('').each_with_index do |char, idx|   # split string into an array and iterate over each element and get the index
    subs << char  # shovel each charachter into the array
    fragment = char   # create a variable to store each character

    string[idx + 1..-1].split('').each_with_index do |letter|   # split the string into array from index 1 to second to last element. Iterate and get each letter
      fragment += letter  # add the letter to the fragment 
      subs << fragment
    end
  end

  subs

end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: ['a', 'ab', 'abc', 'b', 'bc', 'c']"
  print "=> " 
  print consecutive_substrings('abc')

  puts

  puts "Expecting: ['a']"
  print "=> " 
  print consecutive_substrings('a')

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution

File: 03-week-3--additional-practice/04-bonus-5--consecutive-substrings/javascript/consecutive_substrings.js
function consecutiveSubstrings(string) {
  // type your code here
  let result = []
  for (i = 0; i < string.length; i++) {   // first loop for starting index
    // let fragment = string.charAt(i);
    // console.log("frag", fragment);
    // console.log("index", i);
    let substring = ''

      for (let j = i; j < string.length; j++) {   // second loop for generating substring
        substring += string.charAt(j);
        result.push(substring);
      }
      
  }
  return result;
}

function consecutiveSubstringsAlt(string) {
  const stringArr = string.split('');
  const subs = [];

  stringArr.forEach((char, idx) => {
    subs.push(char);

    let fragment = char;

    stringArr.slice(idx + 1).forEach((letter) => {
      fragment = fragment + letter;
      subs.push(fragment);
    });
  });

  return subs;
}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: ['a', 'ab', 'abc', 'b', 'bc', 'c']");
  console.log("=>", consecutiveSubstrings('abc'));

  console.log("");

  console.log("Expecting: ['a']");
  console.log("=>", consecutiveSubstrings('a'));

  console.log("");

  console.log("Expecting: ['j, jo, joh, john, o, oh, ohn, h, hn, n']");
  console.log("=>", consecutiveSubstrings('john'));

  console.log("");

  console.log("Expecting: ['b, bo, bob, o, ob, b']");
  console.log("=>", consecutiveSubstrings('bob'));

  console.log("");

  console.log("Expecting: []");
  console.log("=>", consecutiveSubstrings(''));

  console.log("");

}

function benchmark(callback) {
  const startTime = Date.now();

  for (let i = 0; i < 1000; ++i) {
    callback();
  }

  return (Date.now() - startTime) / 1000;
}

console.log('for loop:', benchmark(() => consecutiveSubstrings('johnogletree')));
console.log('forEach:', benchmark(() => consecutiveSubstringsAlt('johnogletree')));

module.exports = consecutiveSubstrings;

// Please add your pseudocode to this file
// And a written explanation of your solution
